Public/Src/FrontEnd/Nuget/NugetCgManifestGenerator.cs

```diff
@@ -0,0 +1,102 @@
+
+using System.Linq;
+using BuildXL.FrontEnd.Sdk;
+using BuildXL.Utilities.Collections;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace BuildXL.FrontEnd.Nuget
+{
+    /// <summary>
+    /// NugetCgManifestGenerator is used for creation and comparasion of a manifest file for Component Governance.
+    /// The cgmanifest file contains information about all the Nuget Packages used in BuildXL with all their versions in use
+    /// The cgmanifest file is used by Component Governance to determine security risks within components used by BuildXL
+    /// This manifest file will only be picked up for Component Governance if it is named "cgmanifest.json" as per cg documentation: https://docs.opensource.microsoft.com/tools/cg.html
+    /// </summary>
+    public sealed class NugetCgManifestGenerator
+    {
+        private FrontEndContext Context { get; }
+
+        /// <nodoc />
+        public NugetCgManifestGenerator(FrontEndContext context)
+        {
+            Context = context;
+        }
+
+        /// <summary>
+        /// Generates json as an indented string containing all the NuGet package names and versions used in BuildXL
+        /// </summary>
+        public string GenerateCgManifestForPackages(MultiValueDictionary<string, Package> packages)
+        {
+            var components = packages
+                .Keys
+                .SelectMany(nugetName => packages[nugetName].Select(package => new NugetPackageAndVersionStore(nugetName, ExtractNugetVersion(package))))
+                .OrderBy(c => c.Name)
+                .ThenBy(c => c.Version)
+                .Select(c => ToNugetComponent(c.Name, c.Version))
+                .ToList();
+
+            var cgmanifest = new
+            {
+                Version = 1,
+                Registrations = components
+            };
+
+            return JsonConvert.SerializeObject(cgmanifest, Formatting.Indented);
+        }
+
+        /// <summary>
+        /// Compares <paramref name="lhsManifest"/> and <paramref name="rhsManifest"/> for equality;
+        /// returns true if they are equal, and false otherwise. 
+        /// 
+        /// This equality check is case-insensitive and white space agnostic.
+        /// </summary>
+        public bool CompareForEquality(string lhsManifest, string rhsManifest)
+        {
+            try
+            {
+                return JToken.DeepEquals(JObject.Parse(lhsManifest), JObject.Parse(rhsManifest));
+            }
+            catch (JsonReaderException)
+            {
+                // The existing Manifest file was in invalid JSON format.
+                // Hence it does not match.
+                return false;
+            }
+            
+        }
+
+        private string ExtractNugetVersion(Package p)
+        {
+            return p.Path.GetParent(Context.PathTable).GetName(Context.PathTable).ToString(Context.StringTable);

```

Comment:
Let's leave a comment that this relies on the folder structure created by the Nuget resolver.   Ideally, it would be great to find the code that creates that folder structure and cross reference this method in there to say that if that gets changed this method needs to be updated as well.